auth-service/src/main/kotlin/uk/gov/communities/delta/auth/services/MemberOfToDeltaRolesMapper.kt

```diff
@@ -0,0 +1,173 @@
+    }
+
+    private class ParsedRole(val role: String, val orgId: String?, val type: RoleType, val originalGroup: String)
+
+
+    private fun removeDatamartDeltaPrefix(group: String): String? {
+        return if (group.startsWith(DATAMART_DELTA_PREFIX)) {
+            group.substring(DATAMART_DELTA_PREFIX.length, group.length)
+        } else {
+            logger.atWarn().addKeyValue("username", username).addKeyValue("group", group)
+                .log("Ignoring group as CN does not start with $DATAMART_DELTA_PREFIX")
+            null
+        }
+    }
+
+    private fun parseToRoleAndOrg(groupWithoutPrefix: String): ParsedRole? {
+        /*
+         * Group names take the form datamart-delta-<role name>[-<orgId>], e.g.
+         * datamart-delta-data-provider-dclg
+         * ______________-_____________-____
+         *  ^prefix        ^role         ^organisation id
+         * Where role can be a system role (like "data-provider"), or an access group (like "homelessness-research")
+         * The organisation id is optional and indicates the user has that role in that organisation.
+         * Role and organisation ids can both have hyphens in.
+         */
+        val orgId = organisationIdSuffixes.firstOrNull { groupWithoutPrefix.endsWith(it) }?.substring(1)
+        val roleStr =
+            if (orgId == null) groupWithoutPrefix
+            else groupWithoutPrefix.substring(0, groupWithoutPrefix.length - (orgId.length + 1))
+
+        val roleType = determineRoleType(roleStr)
+        return if (roleType == null) {
+            logger.atWarn().addKeyValue("username", username).addKeyValue("role", roleStr)
+                .addKeyValue("orgId", orgId).addKeyValue("group", DATAMART_DELTA_PREFIX + groupWithoutPrefix)
+                .log("Ignoring group as unable to find matching system role or access group")
+            null
+        } else {
+            ParsedRole(roleStr, orgId, roleType, DATAMART_DELTA_PREFIX + groupWithoutPrefix)
+        }
+    }
+
+    private fun determineRoleType(role: String): RoleType? {
+        return if (DELTA_SYSTEM_ROLES.contains(role)) RoleType.SYSTEM
+        else if (DELTA_EXTERNAL_ROLES.contains(role)) RoleType.EXTERNAL
+        else if (allAccessGroupsMap.contains(role)) RoleType.ACCESS_GROUP
+        else null
+    }
+}
+
+val DELTA_SYSTEM_ROLES = hashSetOf(
+    "admin",
+    "read-only-admin",
+    "testers",
+    "data-providers",
+    "data-certifiers",
+    "data-auditors",
+    "billing-data-approvers",
+    "warehouse-users",
+    "dataset-admins",
+    "form-designers",
+    "lead-testers",
+    "user",
+    "data-entry-clerks",
+    "payments-approvers",
+    "payments-reviewers",
+    "sap-team-members",
+    "local-admins",
+    "abatements-approvers",
+    "suspensions-approvers",
+    "write-offs-approvers",
+    "setup-managers",
+    "report-users",
+    "personal-data-owners",
+    "stats-data-certifiers",
+)
+
+val DELTA_EXTERNAL_ROLES = hashSetOf("section-151-officers")
```

Comment:
I think I'm missing some context. What's makes section-151-officers an "external" role and need to be handled differently?

Reply:
I think because it's managed by MarkLogic rather than Delta's Java.. though that's a distinction which doesn't currently matter here, could merge them and just have a comment

Reply:
Just added a comment for now, Delta needs the distinction, though it doesn't do much with this role